### PR TITLE
memrchr: Correct aligned offset computation

### LINKED ISSUE #35967: memrchr uses unaligned loads
Due to an arithmetic error, the fallback memrchr (used for non-linux) in libstd uses unaligned loads, the intention was to produce only well aligned pointers of course.

This was suspected to cause a crash using ARMv7, found by @tomaka 

Original code also had the same issue: BurntSushi/rust-memchr/pull/9

I believe that on x86 non-linux platforms, the bug was harmless and the function still produced the correct return value.

### PR DESCRIPTION
**Pull request by bluss**
The memrchr fallback did not compute the offset correctly. It was
intentioned to land on usize-aligned addresses but did not.
This was suspected to have resulted in a crash on ARMv7!

This bug affected non-linux platforms.

I think like this, if we have a slice with pointer `ptr` and length
`len`, we want to find the last usize-aligned offset in the slice.
The correct computation should be:

For example if ptr = 1 and len = 6, and `size_of::<usize>()` is 4:

```
[ x x x x x x ]
  1 2 3 4 5 6
        ^-- last aligned address at offset 3 from the start.
```

The last aligned address is ptr + len - (ptr + len) % usize_size.

Compute offset from the start as:

offset = len - (ptr + len) % usize_size = 6 - (1 + 6) % 4 = 6 - 3 = 3.

I believe the function's return value was always correct previously, if
the platform supported unaligned addresses.

Fixes #35967

### CODE REVIEW

File: src/libstd/memchr.rs [hunk 1]
```diff
@@ -209,7 +209,7 @@ mod fallback {
         let end_align = (ptr as usize + len) & (usize_bytes - 1);
```
Comment (nagisa) [inline src/libstd/memchr.rs]: Preexisting, but this uses `size_of` to calculate `usize_bytes`, but alignment of usize on a system is not `size_of::<usize>` but rather `align_of::<usize>`? Seems like a pessimisation?

File: src/libstd/memchr.rs [hunk 2]
```diff
@@ -209,7 +209,7 @@ mod fallback {
...
         let mut offset;
         if end_align > 0 {
-            offset = len - cmp::min(usize_bytes - end_align, len);
+            offset = len - cmp::min(end_align, len);
```
Comment (nagisa) [inline src/libstd/memchr.rs]: This seems correct to me, but I think the code would be somewhat clearer as explicit if:
```
offset = if end_align > len { 0 } else { len - end_align };
```

**Comment by nagisa** [comment]
LGTM, but it took me some 10 mins in a distraction-ful environment to figure out the offset calculation. I think a bug like this would’ve been obvious with a plain conditional as I proposed above.

**Comment by bluss** [comment]
Thank you for contributing the concentration! I agree that the conditional is better that way, will amend.

File: src/libstd/memchr.rs (hunk 1 above)
Comment (bluss) [inline src/libstd/memchr.rs]: It doesn't seem like a big deal. I prefer a separate PR to deal with this. Let's have the fix here be minimal, especially if we backport it.

**Comment by bluss** [comment]
```
failures:

---- [incremental] incremental\ich_method_call_trait_scope.rs stdout ----
	thread '[incremental] incremental\ich_method_call_trait_scope.rs' panicked at 'called `Result::unwrap()` on an `Err` value: Error { repr: Os { code: 3, message: "The system cannot find the path specified." } }', C:\bot\slave\auto-win-gnu-32-opt-rustbuild\build\src\libcore\result.rs:783
note: Run with `RUST_BACKTRACE=1` for a backtrace.


failures:
    [incremental] incremental\ich_method_call_trait_scope.rs
```

**Comment by bluss** [comment]
Nominating for beta: It's a crashing bug affecting std::io::stdout (common libstd feature) on a minority platform (discovered on ARMv7 on android).

**Comment by bluss** [comment]
Fix was verified by https://github.com/rust-lang/rust/issues/35967#issuecomment-242996772

**Comment by alexcrichton** [comment]
Libs decided to accept for backport.

**Pull request merged**